src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs

```diff
@@ -578,6 +578,8 @@ public partial class Router : Microsoft.AspNetCore.Components.IComponent, Micros
         public Microsoft.AspNetCore.Components.RenderFragment NotFound { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         [Microsoft.AspNetCore.Components.ParameterAttribute]
         public Microsoft.AspNetCore.Components.EventCallback<Microsoft.AspNetCore.Components.Routing.NavigationContext> OnNavigateAsync { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
+        [Microsoft.AspNetCore.Components.ParameterAttribute]
+        public Microsoft.AspNetCore.Components.RenderFragment<System.Exception> OnNavigateError { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
```

Comment:
It feels that this adds more complexity to the router. If a user were to handle errors within `OnNavigateAsync` why would they choose to use a separate render fragment for it instead of a ```try..catch``` inside their method?

In both cases the user still has to make a gesture for it and at that point I would prefer if we avoid having to add public API for it. The advantage of forcing users to write a try..catch is that all the error handling code is in one place instead of separated between the render fragment and their method handler.

A render fragment makes sense when there is a valuable user action that can be taken, but in this case this is not something you would show to end users as it is likely the result of a programming error and not a user error.

I would suggest that instead of doing this, we simply flow the exception through the render method as `_renderHandle.Render(builder => ExceptionDispatchInfo.Capture(completedTask.Exception).Throw())`

Reply:
I tend to agree. We already have a mechanism for displaying an error UI that other lifecycle-type methods will trigger, and other lifecycle methods don't have specialised systems for their own error UI. It would be nice if `OnNavigateAsync` could feed into the same system, with its existing rules about terminating the circuit.

Developers who want to handle navigation exceptions in a custom way could still use try/catch as above.

@javiercn's suggestion about using `_renderHandle.Render` as a way to dispatch the exception back to the renderer is quite a cheeky and surprising hack but I can't think of any actual reasons why it's not an ideal solution :) Assuming it really works, it's great to be able to do it would adding any new public API anywhere.

So maybe this can be reduced to a one-liner!

Reply:
Thanks for the recommendation, @javiercn! I wasn't familiar with the `ExceptionDispatchInfo` API for rethrowing an exception will preserving the context. That makes a lot more sense.

Reply:
This is done now.